### PR TITLE
ENH: Support SBRef and phase data in ReproIn heuristic

### PR DESCRIPTION
**Pull request by tsalo**
Closes https://github.com/ReproNim/reproin/issues/39.

Changes proposed:
- Remove 'P' from automatic sequence type detection. Originally it assumed all 'P' files are field maps, but they can also be phase data.
- Expand sequence type detection for functional scans to include single-band reference volumes (if the series description ends with `_SBRef`) and functional phase data (if the image type tuple includes one entry that is 'P').
- ~~Fix a couple of random typos. And I guess Atom removed some trailing whitespace.~~
     - Moved into #396.

### CODE REVIEW

File: heudiconv/heuristics/reproin.py [hunk 1]
```diff
@@ -541,8 +540,14 @@ def infotodict(seqinfo):
 
         # analyze s.protocol_name (series_id is based on it) for full name mapping etc
         if seqtype == 'func' and not seqtype_label:
-            if '_pace_' in series_spec:
+            if s.series_description.endswith('_SBRef'):
+                seqtype_label = 'sbref'
```
Comment (tsalo) [inline heudiconv/heuristics/reproin.py]: I am not sure how generalizable this is. Single-band reference scans are automatically outputted in an [X]_SBRef dicom folder with the CMRR MB sequence, where [X] is the name of the multi band scan.

File: heudiconv/heuristics/reproin.py [hunk 2]
```diff
@@ -541,8 +540,14 @@ def infotodict(seqinfo):
...
+            elif '_pace_' in series_spec:
                 seqtype_label = 'pace'  # or should it be part of seq-
+            elif 'P' in s.image_type:
+                seqtype_label = 'phase'
```
Comment (yarikoptic) [inline heudiconv/heuristics/reproin.py]: this is already under `seqtype == 'func'` so if I read the code correctly, it wouldn't work for a simple `fmap`

File: heudiconv/heuristics/reproin.py [hunk 3]
```diff
@@ -483,7 +483,6 @@ def infotodict(seqinfo):
             # 3 - Image IOD specific specialization (optional)
             dcm_image_iod_spec = s.image_type[2]
             image_type_seqtype = {
-                'P': 'fmap',   # phase
```
Comment (yarikoptic) [inline heudiconv/heuristics/reproin.py]: if you could split out typo fixes to a separate commit/PR, I would be happy to accept it really quickly ;)

File: heudiconv/heuristics/reproin.py (hunk 2 above)
Comment (tsalo) [reply heudiconv/heuristics/reproin.py]: I think a separate if statement needs to be added for `seqtype == 'fmap'` to check `if 'P' in s.image_type`. Does that sound right?

Comment (tsalo) [reply heudiconv/heuristics/reproin.py]: Wait, sorry, if `seqtype` is already known to be `fmap` then there's no reason to check `s.image_type`. I don't know if it's a problem though, since there's definitely a step that determines `seqtype_label` for field maps that's separate from this part. So this if statement should only apply to functional data.

File: heudiconv/heuristics/reproin.py (hunk 1 above)
Comment (yarikoptic) [reply heudiconv/heuristics/reproin.py]: note: for me, with whatever sequence we have on our Siemens now (yet to check), `_SBRef` is present only within `SeriesDescription` field, and not `ProtocolName`:
```
$> grep s4p2 000001.txt 
(0008,103e) LO [func_run-01_task-pinel_acq-s4p2-me_SBRef] #  40, 1 SeriesDescription
(0018,1030) LO [func_run-01_task-pinel_acq-s4p2-me]     #  34, 1 ProtocolName
```
and in reproin we "favor" (consider first)  `ProtocolName`:
```shell
$> grep '^series_spec_fields' heudiconv/heuristics/reproin.py
series_spec_fields = ('protocol_name', 'series_description')
```
I have ran into the same situation with other types of data, requiring to switch to use `series_description` in favor of `protocol_name`.  I wonder how safe it would be to switch order generally. Otherwise `reproin` should just become smarter...

Comment (tsalo) [reply heudiconv/heuristics/reproin.py]: Should I open an issue about it in the ReproIn repository?

Comment (yarikoptic) [reply heudiconv/heuristics/reproin.py]: nah -- let's solve it HERE!  Sorry for the delay.  Absence of sbref handling in reproin started to bite me, so thank you for leading this effort here! ;)
I will push some tuneups shortly, but here is my question:  When would we get `_phase` images, in particular for "bold"?  Never saw those and do not see BIDS provisioning for them.  BIDS talks about [fmap "Case 2: Two phase images and two magnitude images"](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#case-2-two-phase-images-and-two-magnitude-images)  but that is something we do not get a grip in reproin -- to review multiple scans not independently from each other to decide on indexes (such as `_phase1` and `_phase2` here)

**Comment by yarikoptic** [comment]
ok, I went a bit wild with my additional changes, but individual commits explain reasoning ;) I think `sbref` handling is special kind -- should just override the seqtype whatever it was decided on.  So I moved it out from decision making. Please correct me if I am wrong. Meanwhile I will do a run on some dataset to verify that old behavior is preserved.

**Comment by yarikoptic** [comment]
eh, more to be done ... running on a sample with MB multi-echo data, 

<details>
<summary>here is a difference from master</summary> 

```shell
$> heudiconv/utils/test-compare-two-versions.sh heudiconv{-master,} --bids -f reproin --files dbic-misc/1075_sequence/sourcedata/sub-sid001567
Not running heudiconv since compare-versions/v0.7.0-1-g57c9c56 already exists
Running /home/yoh/picts/mris/heudiconv/venvs/dev3/bin/heudiconv with log in compare-versions/v0.7.0-12-ga5a8eba.log
Results differ: /home/yoh/picts/mris/compare-versions/v0.7.0-1-g57c9c56-v0.7.0-12-ga5a8eba.diff
 .heudiconv/sid001567/info/filegroup.json                                                      | 2144 +++++++++++++++++++++++++++++++-------------------------------
 .heudiconv/sid001567/info/heuristic.py                                                        |   65 +
 .heudiconv/sid001567/info/sid001567.auto.txt                                                  |    4 
 .heudiconv/sid001567/info/sid001567.edit.txt                                                  |    4 
 sourcedata/sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_bold__dup-01.dicom.tgz | 1985 ----------------------------------------------------------
 sourcedata/sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_sbref.dicom.tgz        | 2033 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 sourcedata/sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_bold__dup-01.dicom.tgz    | 4275 -----------------------------------------------------------------------------------------------------------------------------
 sourcedata/sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_sbref.dicom.tgz           | 4260 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 sub-sid001567/anat/sub-sid001567_acq-MPRAGE_T1w.json                                          |  682 -------------------
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_bold.json                         |  102 --
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_bold__dup-01.json                 |  162 ----
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_bold__dup-01.nii.gz               | 1891 -------------------------------------------------------
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_sbref.json                        |   68 +
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_sbref.nii.gz                      | 1891 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-1_bold__dup-01.json             |  133 ---
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-1_bold__dup-01.nii.gz           | 1389 ----------------------------------------
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-2_bold__dup-01.json             |  133 ---
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-2_bold__dup-01.nii.gz           | 1346 ---------------------------------------
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-3_bold__dup-01.json             |  133 ---
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-3_bold__dup-01.nii.gz           | 1331 --------------------------------------
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_rec-magnitude_echo-1_sbref.json      |  133 +++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_rec-magnitude_echo-1_sbref.nii.gz    | 1389 ++++++++++++++++++++++++++++++++++++++++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_rec-magnitude_echo-2_sbref.json      |  133 +++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_rec-magnitude_echo-2_sbref.nii.gz    | 1346 +++++++++++++++++++++++++++++++++++++++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_rec-magnitude_echo-3_sbref.json      |  133 +++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_rec-magnitude_echo-3_sbref.nii.gz    | 1331 ++++++++++++++++++++++++++++++++++++++
 sub-sid001567/sub-sid001567_scans.tsv                                                         |    8 
 task-pinel1_acq-s8XBoulder_bold.json                                                          |  101 --
 28 files changed, 13842 insertions(+), 14763 deletions(-)

```
</details>

I get those `_rec-magnitude` in multi-echo `_sbref`, which shouldn't be there...

File: heudiconv/heuristics/reproin.py (hunk 1 above)
Comment (tsalo) [reply heudiconv/heuristics/reproin.py]: You get both `_bold` and `_phase` when you do Magnitude+Phase reconstruction. That's also when you get `rec-magnitude` and `rec-phase` sbrefs (as in #421). Phase is listed as one of the valid task MRI contrasts in the BIDS specification [here](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#task-including-resting-state-imaging-data).

**Comment by tsalo** [comment]
> I get those _rec-magnitude in multi-echo _sbref, which shouldn't be there...

I think that is a pre-existing issue, given #422.

**Comment by yarikoptic** [comment]
Thank you @tsalo !  apparently I didn't have a cohesive picture of the situation!  I guess we might just absorb and improve (according to my [comment there](https://github.com/nipy/heudiconv/pull/422#discussion_r382853282)) #422 into this PR to proceed.  I will do it later today, and then redo testing

File: heudiconv/heuristics/reproin.py (hunk 1 above)
Comment (yarikoptic) [reply heudiconv/heuristics/reproin.py]: THANK YOU!  I was looking for `_phase` and thus didn't hit that BIDS spec link you gave.

**Comment by tsalo** [comment]
I don't want to overcomplicate things, but #424 should also address the `rec` issue (see [this comment](https://github.com/nipy/heudiconv/pull/424/files#r397984207)).

**Comment by yarikoptic** [comment]
> I don't want to overcomplicate things, but #424 should also address the `rec` issue (see [this comment](https://github.com/nipy/heudiconv/pull/424/files#r397984207)).

eh, I already overcomplicated this PR ;)  Just pushed with additional tune ups, will do some testing now to see if we got stable grounds and then will review #424

**Comment by yarikoptic** [comment]
hm, for some reasons CI isn't running :-/

**Comment by yarikoptic** [comment]
<details>
<summary>FWIW -- current version within PR seems to do the right thing</summary> 

```shell
$> heudiconv/utils/test-compare-two-versions.sh heudiconv{-master,} --bids -f reproin --files dbic-misc/1075_sequence/sourcedata/sub-sid001567
Not running heudiconv since compare-versions/v0.7.0-1-g57c9c56 already exists
Running /home/yoh/picts/mris/heudiconv/venvs/dev3/bin/heudiconv with log in compare-versions/v0.7.0-18-g22c571d.log
Results differ: /home/yoh/picts/mris/compare-versions/v0.7.0-1-g57c9c56-v0.7.0-18-g22c571d.diff
 .heudiconv/sid001567/info/filegroup.json                                                      | 2144 +++++++++++++++++++++++++++++++-------------------------------
 .heudiconv/sid001567/info/heuristic.py                                                        |   65 +
 .heudiconv/sid001567/info/sid001567.auto.txt                                                  |    4 
 .heudiconv/sid001567/info/sid001567.edit.txt                                                  |    4 
 sourcedata/sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_bold__dup-01.dicom.tgz | 1985 ----------------------------------------------------------
 sourcedata/sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_sbref.dicom.tgz        | 2033 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 sourcedata/sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_bold__dup-01.dicom.tgz    | 4275 -----------------------------------------------------------------------------------------------------------------------------
 sourcedata/sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_sbref.dicom.tgz           | 4260 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_bold__dup-01.json                 |  162 ----
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_bold__dup-01.nii.gz               | 1891 -------------------------------------------------------
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_sbref.json                        |  162 ++++
 sub-sid001567/func/sub-sid001567_task-pinel1_acq-s8XBoulder_sbref.nii.gz                      | 1891 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-1_bold__dup-01.json             |  133 ---
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-1_bold__dup-01.nii.gz           | 1389 ----------------------------------------
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-1_sbref.json                    |  133 +++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-1_sbref.nii.gz                  | 1389 ++++++++++++++++++++++++++++++++++++++++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-2_bold__dup-01.json             |  133 ---
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-2_bold__dup-01.nii.gz           | 1346 ---------------------------------------
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-2_sbref.json                    |  133 +++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-2_sbref.nii.gz                  | 1346 +++++++++++++++++++++++++++++++++++++++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-3_bold__dup-01.json             |  133 ---
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-3_bold__dup-01.nii.gz           | 1331 --------------------------------------
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-3_sbref.json                    |  133 +++
 sub-sid001567/func/sub-sid001567_task-pinel2_acq-s4p2Xme_echo-3_sbref.nii.gz                  | 1331 ++++++++++++++++++++++++++++++++++++++
 sub-sid001567/sub-sid001567_scans.tsv                                                         |    8 
 25 files changed, 13933 insertions(+), 13881 deletions(-)
```
</details>

**Comment by yarikoptic** [comment]
FWIW I am of opinion that #424 is worthwhile on its own to follow this one, which is still relatively not that "revolutionary" ;)   Meanwhile, I will push an amended last commit to hopefully trigger testing, and will do few more regression tests locally

**Comment by yarikoptic** [comment]
ok, adjusted conditioning on the warning... still looking for a sample (small) dataset which we could use to add some basic testing.

**Comment by yarikoptic** [comment]
ok, not sure when/if I find a good dataset to do testing, so - let's merge and then proceed with #424 (I will take a stab at resolving conflicts later today, unless you @tsalo would stop me and take over ;))
